backend/app/services/cluster_service.py

```
@@ -0,0 +1,104 @@
+    # 2. Validate all farmers exist and belong to the claimed village
+    for fid in cluster_data.farmer_ids:
+        farmer = farmer_repo.get(fid)
+        if not farmer:
+            raise NotFoundError(f"Farmer {fid} not found")
+        if farmer.get("village") != cluster_data.village:
+            raise ValidationAppError(f"Farmer {fid} does not belong to village {cluster_data.village}")
+
+    # 3. Verify centre exists
+    centre = centre_repo.get(cluster_data.centre_id)
+    if not centre:
+        raise NotFoundError(f"Centre {cluster_data.centre_id} not found")
```

Ruff (0.16.3)

[warning] Avoid specifying long messages outside the exception class

(TRY003)